scipy/spatial/_ckdtree.pyx

```diff
@@ -550,69 +564,111 @@ cdef class cKDTree:
+            if leafsize<1:
+                raise ValueError("leafsize must be at least 1")
 
-        cself.tree_buffer = new vector[ckdtreenode]()
-
-        tmpmaxes = np.copy(self.maxes)
-        tmpmins = np.copy(self.mins)
-
-        ptmpmaxes = &tmpmaxes[0]
-        ptmpmins = &tmpmins[0]
-        with nogil:
-            build_ckdtree(cself, 0, cself.n, ptmpmaxes, ptmpmins, median, compact)
-
-        # set up the tree structure pointers
-        self._post_init()
+            if boxsize is None:
+                self.boxsize = None
+                self.boxsize_data = None
+            else:
+                self.boxsize_data = np.empty(2 * self.m, dtype=np.float64)
+                boxsize = broadcast_contiguous(boxsize, shape=(self.m,),
+                                            dtype=np.float64)
+                self.boxsize_data[:self.m] = boxsize
+                self.boxsize_data[self.m:] = 0.5 * boxsize
+
+                self.boxsize = boxsize
+                periodic_mask = self.boxsize > 0
+                if ((self.data >= self.boxsize[None, :])[:, periodic_mask]).any():
+                    raise ValueError("Some input data are greater than the size of the periodic box.")
+                if ((self.data < 0)[:, periodic_mask]).any():
+                    raise ValueError("Negative input data are outside of the periodic box.")
+
+            self.maxes = np.ascontiguousarray(
+                np.amax(self.data, axis=0) if self.n > 0 else np.zeros(self.m),
+                dtype=np.float64)
+            self.mins = np.ascontiguousarray(
+                np.amin(self.data,axis=0) if self.n > 0 else np.zeros(self.m),
+                dtype=np.float64)
+            self.indices = np.ascontiguousarray(np.arange(self.n,dtype=np.intp))
+
+            self._pre_init()
+
+            compact = 1 if compact_nodes else 0
+            median = 1 if balanced_tree else 0
+
+            cself.tree_buffer = new vector[ckdtreenode]()
+
+            tmpmaxes = np.copy(self.maxes)
+            tmpmins = np.copy(self.mins)
+
+            ptmpmaxes = &tmpmaxes[0]
+            ptmpmins = &tmpmins[0]
+            with nogil:
+                build_ckdtree(cself, 0, cself.n, ptmpmaxes, ptmpmins, median, compact)
+
+            # set up the tree structure pointers
+            self._post_init()
+            self._init_complete = True
+        finally:
+            self.release_init_lock()
+
+    cdef void acquire_init_lock(self):
+        if not PyThread_acquire_lock(self._init_lock, 0):
+            PyThread_acquire_lock(self._init_lock, 1)
+
+    cdef void release_init_lock(self):
+        PyThread_release_lock(self._init_lock)
+
+    cdef void acquire_tree_lock(self):
+        if not PyThread_acquire_lock(self._tree_lock, 0):
+            PyThread_acquire_lock(self._tree_lock, 1)
+
+    cdef void release_tree_lock(self):
+        PyThread_release_lock(self._tree_lock)
+
+    cdef void check_init(self):
```

Comment:
Get rid of this nonsense. Aquiring and releasing a lock as a memory barrier? We have Event object or atomic variables for that.